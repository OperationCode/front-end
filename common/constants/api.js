import { get, post, patch } from 'common/utils/api-utils';

/* GET REQUESTS */
export const getCodeSchoolsPromise = () => get('code_schools');
export const getMentorPromise = id => get(`mentors/${id}`);
export const getMentorsPromise = () => get('mentors');
export const getRequestsPromise = () => get('requests');
export const getScholarshipPromise = id => get(`scholarships/${id}`);
export const getScholarshipsPromise = () => get('scholarships');
export const getServicesPromise = () => get('services');
export const getTeamMembersPromise = () => get('team_members');

/* POST REQUESTS */
export const createUser = ({ email, password, firstName, lastName, zipcode }) =>
  post('users', {
    user: {
      email,
      password,
      first_name: firstName,
      last_name: lastName,
      zip: zipcode,
    },
  }).then(({ data }) => data);

export const loginUser = ({ email, password }) =>
  post('sessions', {
    user: {
      email,
      password,
    },
  }).then(({ data }) => data);

export const postMentorRequestPromise = ({ language, additionalDetails, mentor, service }) =>
  post('requests', {
    request: {
      details: additionalDetails,
      language,
      requested_mentor_id: mentor,
      service_id: service,
    },
  });

/* PATCH REQUESTS */
export const patchUpdateMentorRequestPromise = ({ request, status, mentor }) =>
  patch(`requests/${request}`, {
    request: {
      status,
      mentor,
    },
  });
