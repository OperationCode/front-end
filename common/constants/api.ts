import { patch, post } from 'common/utils/api-utils';
import type { RegistrationFormValues } from 'components/Forms/RegistrationForm/RegistrationForm';
import type { MilitaryStatusFormShape } from 'components/Forms/UpdateProfileForm/steps/MilitaryStatus';
import type { ProfessionalDetailsFormShape } from 'components/Forms/UpdateProfileForm/steps/ProfessionalDetails';

export const registerUser = (values: RegistrationFormValues) =>
  post('/api/registration/new', values);

export const updateUser = (values: ProfessionalDetailsFormShape | MilitaryStatusFormShape) =>
  patch('/api/registration/update', values);
