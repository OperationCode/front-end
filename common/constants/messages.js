export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailsMatch: 'Emails must match',
  emailExists: 'A user is already registered with this e-mail address',
  password:
    'Minimum password requirements: 8 characters, 1 Lowercase Letter, 1 Uppercase Letter, 1 Number',
  passwordsMatch: 'Passwords must match',
};

export const networkErrorMessages = {
  serverDown: 'Something is wrong on our end. Please try again later.',
};
