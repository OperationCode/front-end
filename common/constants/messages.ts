export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailsMatch: 'Emails must match',
  emailExists: 'A user is already registered with this e-mail address',
  codeOfConduct: 'You must agree to our Code of Conduct',
  slackGuidelines: 'You must agree to our Slack workspace guidelines',
};

export const networkErrorMessages = {
  serverDown: 'Something is wrong on our end. Please try again later.',
};
