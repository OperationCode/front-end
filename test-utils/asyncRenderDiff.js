import { wait } from '@testing-library/react';

/**
 * @description Used to rerender after asynchronous things (Formik forms in particular)
 * @see /components/LoginForm/LoginForm.js for example usage
 * @exports
 * @param {Object.<string, any>} enzymeWrapper
 */
const asyncRenderDiff = async enzymeWrapper => {
  await wait();
  enzymeWrapper.update();
};

export default asyncRenderDiff;
