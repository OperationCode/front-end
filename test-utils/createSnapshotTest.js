/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
