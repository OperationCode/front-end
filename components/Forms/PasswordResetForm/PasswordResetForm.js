import React, { useState } from 'react';
import { string, func, shape } from 'prop-types';
import { Field, Formik } from 'formik';
import * as Yup from 'yup';
import Button from 'components/Buttons/Button/Button';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import { validationErrorMessages } from 'common/constants/messages';
import { getServerErrorMessage } from 'common/utils/api-utils';
import styles from './PasswordResetForm.module.css';

const passwordResetSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
});

PasswordResetForm.propTypes = {
  passwordReset: func.isRequired,
  onSuccess: func.isRequired,
  initialValues: shape({
    email: string,
  }),
};

PasswordResetForm.defaultProps = {
  initialValues: {
    email: '',
  },
};

function PasswordResetForm({ initialValues, onSuccess, passwordReset }) {
  const [errorMessage, setErrorMessage] = useState('');

  const handleSubmit = async ({ email }, actions) => {
    try {
      const { detail } = await passwordReset({ email });
      actions.setSubmitting(false);
      actions.resetForm();

      await onSuccess({ detail });
    } catch (error) {
      actions.setSubmitting(false);

      setErrorMessage(getServerErrorMessage(error));
    }
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleSubmit}
      validationSchema={passwordResetSchema}
    >
      {({ isSubmitting }) => (
        <Form className={styles.PasswordResetForm}>
          <div className={styles.row}>
            <Field
              type="email"
              name="email"
              label="Email*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="username email"
            />
          </div>

          <div className={styles.row}>
            {errorMessage && <Alert type="error">{errorMessage}</Alert>}
          </div>

          <div className={styles.row}>
            <Button
              className={styles.topMargin}
              type="submit"
              theme="secondary"
              disabled={isSubmitting}
            >
              Submit
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

export default PasswordResetForm;
