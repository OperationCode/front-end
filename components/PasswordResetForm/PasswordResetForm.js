import React from 'react';
import { string, func, shape } from 'prop-types';
import { Field, Formik } from 'formik';
import * as Yup from 'yup';
import Button from 'components/Button/Button';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import { validationErrorMessages } from 'common/constants/messages';
import { getServerErrorMessage } from 'common/utils/api-utils';
import styles from './PasswordResetForm.css';

const passwordResetSchema = Yup.object().shape({
  email: Yup.string()
    .required(validationErrorMessages.required)
    .email(validationErrorMessages.email),
});

export default class PasswordResetForm extends React.Component {
  static propTypes = {
    passwordReset: func.isRequired,
    onSuccess: func.isRequired,
    initialValues: shape({
      email: string,
    }),
  };

  static defaultProps = {
    initialValues: {
      email: '',
    },
  };

  state = {
    errorMessage: '',
  };

  handleSubmit = async ({ email }, actions) => {
    const { passwordReset, onSuccess } = this.props;
    try {
      const { detail } = await passwordReset({ email });
      actions.setSubmitting(false);
      actions.resetForm();

      await onSuccess({ detail });
    } catch (error) {
      actions.setSubmitting(false);

      this.setState({ errorMessage: getServerErrorMessage(error) });
    }
  };

  render() {
    const { props, state } = this;

    return (
      <Formik
        initialValues={props.initialValues}
        onSubmit={this.handleSubmit}
        validationSchema={passwordResetSchema}
      >
        {({ isSubmitting }) => (
          <Form className={styles.PasswordResetForm}>
            <div className={styles.row}>
              <Field
                type="email"
                name="email"
                label="Email*"
                component={Input}
                disabled={isSubmitting}
                autoComplete="username email"
              />
            </div>

            <div className={styles.row}>
              {state.errorMessage && <Alert type="error">{state.errorMessage}</Alert>}
            </div>

            <div className={styles.row}>
              <Button
                className={styles.topMargin}
                type="submit"
                theme="secondary"
                disabled={isSubmitting}
              >
                Submit
              </Button>
            </div>
          </Form>
        )}
      </Formik>
    );
  }
}
