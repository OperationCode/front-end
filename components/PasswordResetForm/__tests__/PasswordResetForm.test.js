import React from 'react';
import { mount } from 'enzyme';
import { wait } from '@testing-library/react';
import { passwordReset } from 'common/constants/api';
import { validationErrorMessages } from 'common/constants/messages';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import mockUser from 'test-utils/mockGenerators/mockUser';
import asyncRenderDiff from 'test-utils/asyncRenderDiff';
import PasswordResetForm from '../PasswordResetForm';

afterEach(() => {
  OperationCodeAPIMock.reset();
});

describe('PasswordResetForm', () => {
  it('should render with required props', () => {
    createSnapshotTest(<PasswordResetForm onSuccess={jest.fn()} passwordReset={jest.fn()} />);
  });

  it('should display required error message when blurring past email input', async () => {
    const wrapper = mount(<PasswordResetForm onSuccess={jest.fn()} passwordReset={jest.fn()} />);

    wrapper.find('input#email').simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="email"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.required);
  });

  it('should show error when providing non-email to email input', async () => {
    const wrapper = mount(<PasswordResetForm onSuccess={jest.fn()} passwordReset={jest.fn()} />);
    wrapper
      .find('input#email')
      .simulate('change', { target: { id: 'email', value: 'email' } })
      .simulate('blur');

    await asyncRenderDiff(wrapper);

    expect(
      wrapper
        .find('Input[type="email"]')
        .find('Alert')
        .text(),
    ).toStrictEqual(validationErrorMessages.email);
  });

  it('should submit with valid data in form', async () => {
    const user = mockUser();

    OperationCodeAPIMock.onPost('auth/password/reset/', { email: user.email }).reply(200, {
      detail: 'success',
    });

    const successSpy = jest.fn();

    const wrapper = mount(
      <PasswordResetForm onSuccess={successSpy} passwordReset={passwordReset} />,
    );

    wrapper
      .find('input#email')
      .simulate('change', { target: { id: 'email', value: user.email } })
      .simulate('blur');

    wrapper.find('Button').simulate('submit');

    await asyncRenderDiff(wrapper);

    await wait(() => {
      expect(successSpy).toHaveBeenCalledWith({ detail: 'success' });
      expect(OperationCodeAPIMock.history.post.length).toBeGreaterThan(0);
    });
  });

  it('should display an error message if request fails', async () => {
    const user = mockUser();
    OperationCodeAPIMock.onPost('auth/password/reset/', { email: user.email }).reply(400, {
      error: 'test error',
    });

    const successSpy = jest.fn();

    const wrapper = mount(
      <PasswordResetForm onSuccess={successSpy} passwordReset={passwordReset} />,
    );

    wrapper
      .find('input#email')
      .simulate('change', { target: { id: 'email', value: user.email } })
      .simulate('blur');

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    expect(wrapper.find('Alert').text()).toStrictEqual('test error');
  });

  it('should NOT submit with invalid data in form', async () => {
    const initialValues = {
      email: 'email',
    };

    const successSpy = jest.fn();
    const passwordResetSpy = jest.fn();

    const wrapper = mount(
      <PasswordResetForm
        onSuccess={successSpy}
        passwordReset={passwordResetSpy}
        {...initialValues}
      />,
    );

    wrapper.find('Button').simulate('submit');
    await asyncRenderDiff(wrapper);

    await wait(() => {
      expect(passwordResetSpy).not.toHaveBeenCalled();
      expect(successSpy).not.toHaveBeenCalled();
      expect(OperationCodeAPIMock.history.post.length).not.toBeGreaterThan(0);
    });
  });
});
