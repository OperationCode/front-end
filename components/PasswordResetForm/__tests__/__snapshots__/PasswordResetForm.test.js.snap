// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordResetForm should render with required props 1`] = `
<form
  className="PasswordResetForm"
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="Label"
        data-testid="LABEL"
        htmlFor="email"
      >
        Email*
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          autoComplete="username email"
          className="Input"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          type="email"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    
  </div>
  <div
    className="row"
  >
    <button
      className="Button topMargin secondary"
      data-testid="BUTTON"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
